### PR TITLE
Fix for issue #22

### PR DESCRIPTION
**Pull request by ghost**
What these changes do is remove all attempts to recreate the spawning logic and leave that to the superclass, which follows the standard minecraft logic for mob spawning.

The overrides of getBlockPathWeight() and isValidLightLevel() are required for the ENDERMAN_DAY_SPAWN config option to work. Both are defaulted to values which are expected for mobs that spawn on daylight.

I think its better to rely on the entity superclass to resolve which are the normal conditions for spawning rather than to try to rewrite the spawning logic.

### CODE REVIEW

**Comment by p455w0rd** [comment]
Did you test this? If you have Discord, please join: https://discord.gg/9J7yhUP

**Comment by ghost** [comment]
Tested it as best as I could, and I'm going to be using a version of your mod with these changes in my current playthrough of Modern Skyblock 3 (it was what prompted me look into this bug, the endermen are really really annoying). I'll report back if there are any unexpected consequences.

**Comment by p455w0rd** [comment]
cool..i'll keep it open and test myself soon..if everything checks out, I'll merge..thank you! I'm not even sure what `getBlockPathWeight` does (i'd assume it has to do with what's traditionally known as "botpathing"), so I'll look into that method and see how it affects spawning as well.

**Comment by ghost** [comment]
`getBlockPathWeight` has to be a misnomer. It is used by `getCanSpawnHere` to determine if an entity can spawn on a block (in `EntityMob` the method uses light level for that), and is used by the mob AI not in pathfinding but to find shelter from the sun. Maybe it was used once with another purpose and the name was never changed.

**Comment by ghost** [comment]
I did some testing on the spawn rate of evolved enderman with these changes applied, using command blocks to count entities which just this mod loaded. Here are my findings.

---

TL;DR: These changes affect the rate at how evolved enderman spawn, which was to be expected. Maybe it needs to be increased somehow so that they aren't that rare? And maybe find a way to decrease the rate during daylight? Also frienderman seem to be spawning too fast and choking other mob's spawning rate (evolved enderman in particular).

---

With the default config (evolved enderman spawning on daylight):

On overworld during day, evolved enderman population is quite dense (possibly due to the fact that the list of spawnables mobs is a lot shorter during day):



On overworld during night, evolved enderman are rare:



On the nether, evolved enderman are a bit more common than in overworld at night, and frienderman also start off as rare as evolved endermen, but given time friendermen spawn more rapidly and they start to "choke" the spawning of other mobs:





On the end, evolved endermen are pretty common but something similar happens with friendermen, their population increases overtimes and chokes the spawning of other mobs:





With the config changed to stop evolved endermen spawning on daylight, the stats are similar but without spawning while it's day.

**Comment by p455w0rd** [comment]
cool...as far as Frienderman, they're neutral (same as polar bears, but only permitted to naturally spawn in the end and nether). In the End, there are supposed to be lots of them. In the Nether, it's probably due to the fact that most mobs are hostile there. I can likely change spawning to match Pigmen though using the methods you used on the Evolved Enderman and extending whatever class Pigmen extend. Just ideas. If you feel you can further improve the PR, I'm all ears :D

**Comment by ghost** [comment]
Then... from what your telling me... Friendermen spawning mechanics are as expected, they should be more common than Evolved Enderman in both the End and the Nether.

What worries me is how these proposed changes affect Evolved Enderman population. I'm going to run the same test on the master branch and compare the results.

**Comment by ghost** [comment]
Well, I ran a few tests on the master branch and the findings are pretty similar. I daresay these changes did not affect expected evolved enderman population. In both cases I think they are spawning too much during day, I have some ideas as to how to solve that, but I don't think this PR is the place for that, the point here was just to fix the issue at hand, and I'm confident it's been solved.

I'm still playing on the Modern Skyblock 3 modpack with a modified jar and I'll report back anything weird that happens, but nothing yet. I'm enjoying skyblock life without pesky enderman stealing my stuff (before you ask, yes, I need them to leave my cactus farm alone, I'm eating that shit). Although I consider this done and I'm moving now to an Agricraft bug that's been bothering me.

Here are some screenshots from testing the population on master branch:

Overworld day:



Overworld night:



Nether:



End:

**Comment by p455w0rd** [comment]
thank you, sorry this took so long :D

**Comment by ghost** [comment]
No prob, relatively this is actually not that long, the project I'm trying to salvage at work it's almost a year late on its due date... I would like to say this is an exception but, in my experience in enterprise level software development, this kind of things happens more than it should. Thank you!
